### PR TITLE
LM-56: implement changes to project controller service dto viewmodel

### PR DESCRIPTION
**Pull request by rgillen-uno**
Changes for Project Controller/Service/Dto/ViewModel.

There may be additional functionality required of the Controller and Service that isn't included currently.

### CODE REVIEW

File: AdminCore.Common/Interfaces/IProjectService.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+namespace AdminCore.Common.Interfaces
+{
+    public interface IProjectService
+    {
+        ProjectDto CreateProject(ProjectDto projectToCreate);
+        ProjectDto UpdateProject(ProjectDto projectToUpdate);
+        void DeleteProject(int projectId);
+
+        IList<ProjectDto> GetProjects();
```
Comment (BrownMark) [inline AdminCore.Common/Interfaces/IProjectService.cs]: Optional paramters could work here.

File: AdminCore.Services.Tests/ProjectServiceTests.cs [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+    private void UpdateProject_ValidProjectDto_AllCallsReceived(ProjectDto projectToUpdate, Project dbReturns)
+    {
+      // Arrange
+      GetMockedResourcesProjectRepo(dbReturns, out var mapper, out var projectService, out var ormContext);
+
+      // Act
+      projectService.UpdateProject(projectToUpdate);
+
+      // Assert
+      ormContext.Received(1).ProjectRepository.Update(dbReturns);
```
Comment (BrownMark) [inline AdminCore.Services.Tests/ProjectServiceTests.cs]: We need a way to verify that this contains the mapped changed.

File: AdminCore.Services/ProjectService.cs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            var projectList = DatabaseContext.ProjectRepository.Get(project => project.ProjectId == projectId, null,
+                project => project.Client,
+                project => project.Teams,
+                project => project.ParentProject);
+            return _mapper.Map<IList<ProjectDto>>(projectList ?? new List<Project>());
+        }
+
+        public IList<ProjectDto> GetProjectsByClientId(int clientId)
+        {
+            var projectList = DatabaseContext.ProjectRepository.Get(project => project.ClientId == clientId, null,
```
Comment (BrownMark) [inline AdminCore.Services/ProjectService.cs]: I think we can consolidate these.

File: AdminCore.WebApi.Tests/Controllers/ProjectControllerTests.cs [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+    {
+      // Arrange
+      GetMockedResourcesGetProjects(null, new List<ProjectViewModel>(), out var projectServiceMock, out var projectController, out var mapper);
+
+      // Act
+      var response = projectController.GetProjects();
+
+      // Assert
+      var result = RetrieveValueFromActionResult<IList<ProjectViewModel>>(response);
+      result.Should().NotBeNull().And.HaveCount(0);
```
Comment (BrownMark) [inline AdminCore.WebApi.Tests/Controllers/ProjectControllerTests.cs]: Could do BeEmpty here.

File: AdminCore.Services/ProjectService.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+                project => project.Teams,
+                project => project.ParentProject);
```
Comment (rgillen-uno) [inline AdminCore.Services/ProjectService.cs]: Include Client?

File: AdminCore.WebApp/Controllers/ProjectController.cs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        return BadRequest();
+      }
+    }
+
+    [HttpPut]
+    [ProducesResponseType(typeof(IList<ProjectViewModel>), StatusCodes.Status200OK)]
+    [ProducesResponseType(StatusCodes.Status400BadRequest)]
+    public IActionResult UpdateProject([FromBody] UpdateProjectViewModel projectToUpdate)
+    {
+      try
```
Comment (BrownMark) [inline AdminCore.WebApp/Controllers/ProjectController.cs:60]: private IActionResult HandleBarRequest(Action updateAction) {
  try {
   action();
  }
  catch (DbUpdateException e)
  {
        return BadRequest();
  }
}

**Review by BrownMark — CHANGES_REQUESTED** [review]

File: AdminCore.Common/Interfaces/IProjectService.cs (hunk 1 above)
Comment (rgillen-uno) [reply AdminCore.Common/Interfaces/IProjectService.cs]: Optional parameters implemented

File: AdminCore.Services/ProjectService.cs (hunk 3 above)
Comment (rgillen-uno) [reply AdminCore.Services/ProjectService.cs]: I think I consolidated these as part of the optional parameter comment?

File: AdminCore.Services.Tests/ProjectServiceTests.cs (hunk 2 above)
Comment (rgillen-uno) [reply AdminCore.Services.Tests/ProjectServiceTests.cs]: Added more specific spies i.e
```
ormContext.Received(1).ProjectRepository.Update(Arg.Is<Project>(x => x.ProjectId == projectToUpdate.ProjectId));
```

Comment (rgillen-uno) [reply AdminCore.Services.Tests/ProjectServiceTests.cs]: Didn't have access to the mapped object before .Update() so compared the Update data with what the spy received.

File: AdminCore.WebApp/Controllers/ProjectController.cs (hunk 6 above)
Comment (rgillen-uno) [reply AdminCore.WebApp/Controllers/ProjectController.cs:60]: Now using this Func<>

File: AdminCore.WebApi.Tests/Controllers/ProjectControllerTests.cs (hunk 4 above)
Comment (rgillen-uno) [reply AdminCore.WebApi.Tests/Controllers/ProjectControllerTests.cs]: Now using this,

**Pull request merged**